automap/src/comm_layer/igdp.rs

```diff
@@ -0,0 +1,1687 @@
+        }
+        let change_log_arc = Arc::new(Mutex::new(vec![]));
+        let inner_arc = change_log_arc.clone();
+        let change_handler =
+            Box::new(move |change: AutomapChange| inner_arc.lock().unwrap().push(change));
+
+        subject
+            .start_housekeeping_thread(change_handler, router_ip)
+            .unwrap();
+
+        thread::sleep(Duration::from_millis(100));
+        let change_log = change_log_arc.lock().unwrap();
+        assert_eq!(
+            *change_log,
+            vec![AutomapChange::Error(AutomapError::CantFindDefaultGateway)]
+        );
+        let inner = subject.inner_arc.lock().unwrap();
+        assert_eq!(inner.public_ip_opt, Some(public_ip));
+        assert!(inner.housekeeping_commander_opt.is_none());
+        TestLogHandler::new().exists_log_containing("ERROR: IgdpTransactor: Can't find router");
+    }
+
+    #[test]
+    fn stop_housekeeping_thread_returns_same_change_handler_sent_into_start_housekeeping_thread() {
+        let change_log_arc = Arc::new(Mutex::new(vec![]));
+        let inner_cla = change_log_arc.clone();
+        let change_handler = Box::new(move |change| {
+            let mut change_log = inner_cla.lock().unwrap();
+            change_log.push(change)
+        });
+        let mut subject = IgdpTransactor::new();
+        subject.public_ip_poll_delay_ms = 10;
+        let _ =
+            subject.start_housekeeping_thread(change_handler, IpAddr::from_str("1.2.3.4").unwrap());
+
+        let change_handler = subject.stop_housekeeping_thread();
+
+        let change = AutomapChange::NewIp(IpAddr::from_str("4.3.2.1").unwrap());
+        change_handler(change.clone());
+        let change_log = change_log_arc.lock().unwrap();
+        assert_eq!(change_log.last().unwrap(), &change)
+    }
+
+    #[test]
+    #[should_panic(expected = "No HousekeepingCommander: can't stop housekeeping thread")]
+    fn stop_housekeeping_thread_handles_missing_housekeeper_commander() {
+        let mut subject = IgdpTransactor::new();
+        subject.inner_arc.lock().unwrap().housekeeping_commander_opt = None;
+
+        let _ = subject.stop_housekeeping_thread();
+    }
+
+    #[test]
+    fn stop_housekeeping_thread_handles_broken_commander_connection() {
+        init_test_logging();
+        let mut subject = IgdpTransactor::new();
+        let (tx, rx) = unbounded();
+        subject.inner_arc.lock().unwrap().housekeeping_commander_opt = Some(tx);
+        std::mem::drop(rx);
+
+        let change_handler = subject.stop_housekeeping_thread();
+
+        change_handler(AutomapChange::Error(AutomapError::NoLocalIpAddress));
+        let tlh = TestLogHandler::new();
+        tlh.exists_log_containing("WARN: IgdpTransactor: Tried to stop housekeeping thread that had already disconnected from the commander");
+        tlh.exists_log_containing("ERROR: IgdpTransactor: Change handler recovery failed: discarded Error(NoLocalIpAddress)");
```

Comment:
I'm unsure about the second log. I couldn't find the place where we would write it in a call of stop_housekeeping_thread().

Ups, sorry, I found it in the null change handler, but it's been really placed unexpectedly. Hopefully null_handler may not need to exist longer.